File: examples/07-aws-containers/java/sqs_signal_bridge.java
// An "EventBridge rule -> Lambda -> StartExecution" trigger becomes a long-poll
// SQS consumer that starts or signals Workflows. The bridge is plain glue code
// (it runs outside any Workflow): receive a message, translate it to a
// signalWithStart, delete it. signalWithStart is idempotent on the Workflow ID,
// so an at-least-once SQS redelivery just re-signals the same Workflow.
class SqsSignalBridge {
  private final WorkflowClient client;
  private final SqsClient sqs;
  private final String queueUrl;

  void pump() {
    while (true) {
      ReceiveMessageResponse resp =
          sqs.receiveMessage(
              ReceiveMessageRequest.builder()
                  .queueUrl(queueUrl)
                  .maxNumberOfMessages(10)
                  .waitTimeSeconds(20) // long poll, not a hot spin
                  .build());

      for (Message m : resp.messages()) {
        FileEvent event = parse(m.body());

        // signalWithStart: starts the Workflow if absent, signals it if running.
        ImportWorkflow stub =
            client.newWorkflowStub(
                ImportWorkflow.class,
                WorkflowOptions.newBuilder()
                    .setWorkflowId("import-" + event.bucket() + "-" + event.key())
                    .setTaskQueue("transform")
                    .build());
        BatchRequest batch = client.newSignalWithStartRequest();
        batch.add(stub::run, event.s3Uri());
        client.signalWithStart(batch);

        // Delete only after the signal is durable in Temporal — at-least-once.
        sqs.deleteMessage(
            DeleteMessageRequest.builder()
                .queueUrl(queueUrl)
                .receiptHandle(m.receiptHandle())
                .build());
      }
    }
  }
}

File: examples/07-aws-containers/python/sqs_signal_bridge.py
"""An "EventBridge rule -> Lambda -> StartExecution" trigger becomes a long-poll
SQS consumer that starts or signals Workflows. The bridge is plain glue code (it
runs outside any Workflow): receive a message, translate it to a start_workflow
(start_signal makes it signal-with-start), delete it. Idempotent on the Workflow
ID, so an at-least-once SQS redelivery just re-signals the same Workflow.

boto3 is imported lazily so this module py_compiles without the AWS SDK.
"""

import json

from temporalio.client import Client


def _sqs_client():
    import boto3  # lazy: AWS SDK may be absent offline

    return boto3.client("sqs")


async def pump(client: Client, queue_url: str) -> None:
    sqs = _sqs_client()
    while True:
        resp = sqs.receive_message(
            QueueUrl=queue_url,
            MaxNumberOfMessages=10,
            WaitTimeSeconds=20,  # long poll, not a hot spin
        )
        for m in resp.get("Messages", []):
            event = json.loads(m["Body"])
            s3_uri = event["s3Uri"]

            # start_signal: starts the Workflow if absent, signals it if running.
            await client.start_workflow(
                "ImportWorkflow",
                s3_uri,
                id=f"import-{event['bucket']}-{event['key']}",
                task_queue="transform",
                start_signal="file_arrived",
                start_signal_args=[s3_uri],
            )

            # Delete only after the signal is durable in Temporal — at-least-once.
            sqs.delete_message(QueueUrl=queue_url, ReceiptHandle=m["ReceiptHandle"])
